kernel/memory_structs/src/lib.rs

```diff
@@ -266,87 +315,145 @@ macro_rules! implement_page_frame {
+                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs * P::CONTAINED_4K)),
+                        size: self.size,
                     };
                 }
             }
-            impl Sub<usize> for $TypeName {
-                type Output = $TypeName;
-                fn sub(self, rhs: usize) -> $TypeName {
+            impl<P: PageSize> Sub<usize> for $TypeName<P> {
+                type Output = $TypeName<P>;
+                fn sub(self, rhs: usize) -> $TypeName<P> {
                     $TypeName {
-                        number: self.number.saturating_sub(rhs),
+                        number: self.number.saturating_sub(rhs * P::CONTAINED_4K),
+                        size: self.size,
                     }
                 }
             }
             impl SubAssign<usize> for $TypeName {
                 fn sub_assign(&mut self, rhs: usize) {
                     *self = $TypeName {
                         number: self.number.saturating_sub(rhs),
+                        size: self.size,
                     };
                 }
             }
-            #[doc = "Implementing `Step` allows `" $TypeName "` to be used in an [`Iterator`]."]
-            impl Step for $TypeName {
+            impl<P: PageSize> Step for $TypeName<P> {
                 #[inline]
-                fn steps_between(start: &$TypeName, end: &$TypeName) -> Option<usize> {
-                    Step::steps_between(&start.number, &end.number)
+                fn steps_between(start: &$TypeName<P>, end: &$TypeName<P>) -> Option<usize> {
+                    match Step::steps_between(&start.number, &end.number) {
+                        None => {
+                            None
+                        }
+                        Some(n) => {
+                            Some(n / P::CONTAINED_4K)
+                        }
+                    }
                 }
                 #[inline]
-                fn forward_checked(start: $TypeName, count: usize) -> Option<$TypeName> {
-                    Step::forward_checked(start.number, count).map(|n| $TypeName { number: n })
+                fn forward_checked(start: $TypeName<P>, count: usize) -> Option<$TypeName<P>> {
+                    Step::forward_checked(start.number, count +  P::CONTAINED_4K).map(|n| $TypeName { number: n, size: PhantomData })
                 }
                 #[inline]
-                fn backward_checked(start: $TypeName, count: usize) -> Option<$TypeName> {
-                    Step::backward_checked(start.number, count).map(|n| $TypeName { number: n })
+                fn backward_checked(start: $TypeName<P>, count: usize) -> Option<$TypeName<P>> {
+                    Step::backward_checked(start.number, count +  P::CONTAINED_4K).map(|n| $TypeName { number: n, size: PhantomData })
+                }
+            }
+            impl TryFrom<&$TypeName> for $TypeName<Page2MiB> {
```

Comment:
`TryFrom` is a "consuming" function, meaning that it should be implemented for `Page`, not `&Page`.